shim/shim.c

```diff
@@ -0,0 +1,654 @@
+	char *buf = NULL;
+	ssize_t need_read = STREAM_HEADER_SIZE;
+	ssize_t bytes_read = 0, want, ret;
+
+	if (! (shim && seq && stream_len)) {
+		return NULL;
+	}
+
+	*stream_len = 0;
+
+	buf = calloc(STREAM_HEADER_SIZE, 1);
+	if (! buf ) {
+		abort();
+	}
+
+	while (bytes_read < need_read) {
+		want = need_read - bytes_read;
+		if (want > BUFSIZ)  {
+			want = BUFSIZ;
+		}
+
+		ret = read(shim->proxy_io_fd, buf+bytes_read, (size_t)want);
+		if (ret == -1) {
+			shim_warning("Error reading from proxy I/O fd: %s\n", strerror(errno));
+			free(buf);
+			return NULL;
+		} else if (ret == 0) {
+			/* EOF received on proxy I/O fd*/
+			shim_warning("EOF received on proxy I/O fd\n");
+			free(buf);
+			return NULL;
+		}
+
+		bytes_read += ret;
+
+		if (*stream_len == 0 && bytes_read >=12) {
+			*stream_len = get_big_endian_32(buf+STREAM_HEADER_LENGTH_OFFSET);
+
+			// length is 12 when hyperstart sends eof before sending exit code
+			if (*stream_len == STREAM_HEADER_SIZE) {
+				break;
+			}
+
+			if (*stream_len > STREAM_HEADER_SIZE) {
+				need_read = *stream_len;
+				buf = realloc(buf, (size_t)*stream_len);
+				if (! buf) {
+					abort();
+				}
+			}
+		}
+	}
+	*seq = get_big_endian_64(buf);
+	return buf;
+}
+
+/*!
+ * Handle output on the proxy I/O fd
+ *
+ *\param shim \ref cc_shim
+ */
+void
+handle_proxy_output(struct cc_shim *shim)
+{
+	uint64_t  seq;
+	char     *buf;
+	int       outfd;
+	ssize_t   stream_len = 0;
+	ssize_t   ret;
+	ssize_t   offset;
+	int       code = 0;
+
+	if (shim == NULL) {
+		return;
+	}
+
+	buf = read_IO_message(shim, &seq, &stream_len);
+	if ( !buf) {
+		/*TODO: is exiting here more appropriate, since this denotes
+		 * error communicating with proxy or proxy has exited
+		 */
+		return;
+	}
+
+	if (seq == shim->io_seq_no) {
+		outfd = STDOUT_FILENO;
+	} else if (seq == shim->io_seq_no + 1) {//proxy allocates errseq 1 higher
+		outfd = STDERR_FILENO;
+	} else {
+		shim_warning("Seq no %"PRIu64 " received from proxy does not match with\
+				 shim seq %"PRIu64 "\n", seq, shim->io_seq_no);
+		free(buf);
+		return;
+	}
+
+	if (!shim->exiting && stream_len == STREAM_HEADER_SIZE) {
+		shim->exiting = true;
+		free(buf);
+		return;
+	} else if (shim->exiting && stream_len == (STREAM_HEADER_SIZE+1)) {
+		code = atoi(buf + STREAM_HEADER_SIZE); 	// hyperstart has sent the exit status
+		free(buf);
+		exit(code);
+	}
+
+	/* TODO: what if writing to stdout/err blocks? Add this to the poll loop
+	 * to watch out for EPOLLOUT
+	 */
+	offset = STREAM_HEADER_SIZE;
+	while (offset < stream_len) {
+		ret = write(outfd, buf+offset, (size_t)(stream_len-offset));
+		if (ret <= 0 ) {
+			free(buf);
+			return;
+		}
+		offset += (ssize_t)ret;
+	}
+	free(buf);
+}
+
+/*!
+ * Handle data on the proxy ctl socket fd
+ *
+ *\param shim \ref cc_shim
+ */
+void
+handle_proxy_ctl(struct cc_shim *shim)
+{
+	char buf[LINE_MAX];
+	ssize_t ret;
+
+	if (! shim) {
+		return;
+	}
+
+	ret = read(shim->proxy_sock_fd, buf, LINE_MAX);
```

Comment:
Maybe `LINE_MAX-1` and initialise `buf` to all zeros for safety: see comment below.